proxy/src/telemetry/metrics/mod.rs

```diff
@@ -597,14 +627,107 @@ impl HyperService for Serve {
                 .with_status(StatusCode::NotFound));
         }
 
-        let body = {
-            let metrics = self.metrics.lock()
-                .expect("metrics lock poisoned");
-            format!("{}", *metrics)
+        let metrics = self.metrics.lock()
+            .expect("metrics lock poisoned")
+            .deref()
+            .clone();
+
+        let is_gzip = is_gzip(&req);
+        let (writer, body) = WriteBody::new();
+        let work = future::lazy(move || {
+            if is_gzip {
+                let mut writer = GzEncoder::new(Vec::<u8>::new(), Compression::Default);
+                write!(writer, "{}", metrics)
+                    .and_then(move |()| {
+                        trace!("gzipping metrics");
+                        writer.finish()
+                    })
+            } else {
+                let mut writer = Vec::<u8>::new();
+                write!(writer, "{}", metrics).map(|_| writer)
+            }
+        })
+        .and_then(|text| tokio_io::io::write_all(writer, text))
+        .and_then(|(w, _)| tokio_io::io::shutdown(w))
+        .map(|_| ())
+        .map_err(|e| {
+            error!("error writing metrics: {:?}", e);
+        });
+
+        self.handle.spawn(work);
+        let encoding = if is_gzip {
+            Encoding::Gzip
+        } else {
+            Encoding::Identity
         };
+
         future::ok(HyperResponse::new()
-            .with_header(ContentLength(body.len() as u64))
+            .with_header(ContentEncoding(vec![encoding]))
```

Comment:
Let's not add the `Content-Encoding` header field unless we're doing a non-identity encoding. Some clients can't handle `identity`.